programs/drift/src/state/lp_pool.rs

```diff
@@ -0,0 +1,1462 @@
+        spot_market: &SpotMarket,
+        price: i64,
+        aum: u128,
+    ) -> DriftResult<i64> {
+        validate!(
+            constituent_index < self.len() as u16,
+            ErrorCode::InvalidConstituent,
+            "Invalid constituent_index = {}, ConstituentTargetBase len = {}",
+            constituent_index,
+            self.len()
+        )?;
+
+        // TODO: validate spot market
+        let datum = self.get(constituent_index as u32);
+        let target_weight = calculate_target_weight(
+            datum.target_base,
+            &spot_market,
+            price,
+            aum,
+            WeightValidationFlags::NONE,
+        )?;
+        Ok(target_weight)
+    }
+}
+
+pub fn calculate_target_weight(
+    target_base: i64,
+    spot_market: &SpotMarket,
+    price: i64,
+    lp_pool_aum: u128,
+    validation_flags: WeightValidationFlags,
+) -> DriftResult<i64> {
+    if lp_pool_aum == 0 {
+        return Ok(0);
+    }
+    let notional: i128 = (target_base as i128)
+        .safe_mul(price as i128)?
+        .safe_div(10_i128.pow(spot_market.decimals))?;
+
+    let target_weight = notional
+        .safe_mul(PERCENTAGE_PRECISION_I128)?
+        .safe_div(lp_pool_aum.cast::<i128>()?)?
+        .cast::<i64>()?
+        .clamp(-1 * PERCENTAGE_PRECISION_I64, PERCENTAGE_PRECISION_I64);
+
+    // if (validation_flags as u8 & (WeightValidationFlags::NoNegativeWeights as u8) != 0)
+    //     && target_weight < 0
+    // {
+    //     return Err(ErrorCode::DefaultError);
+    // }
+    // if (validation_flags as u8 & (WeightValidationFlags::NoOverweight as u8) != 0)
+    //     && target_weight > PERCENTAGE_PRECISION_I64 as i128
+    // {
+    //     return Err(ErrorCode::DefaultError);
+    // }
+
+    // if (validation_flags as u8) & WeightValidationFlags::EnforceTotalWeight100 as u8 != 0 {
+    //     let deviation = (total_weight - PERCENTAGE_PRECISION_I128).abs();
+    //     let tolerance = 100;
+    //     if deviation > tolerance {
+    //         return Err(ErrorCode::DefaultError);
+    //     }
+    // }
+
+    Ok(target_weight)
+}
+
+/// Update target base based on amm_inventory and mapping
+impl<'a> AccountZeroCopyMut<'a, TargetsDatum, ConstituentTargetBaseFixed> {
+    pub fn update_target_base(
+        &mut self,
+        mapping: &AccountZeroCopy<'a, AmmConstituentDatum, AmmConstituentMappingFixed>,
+        // (perp market index, inventory, price)
+        amm_inventory_and_prices: &[(u16, i64, i64)],
+        constituents_indexes_and_decimals_and_prices: &[(u16, u8, i64)],
+        slot: u64,
+    ) -> DriftResult<Vec<i128>> {
+        let mut results = Vec::with_capacity(constituents_indexes_and_decimals_and_prices.len());
+        for (i, constituent_index_and_price) in constituents_indexes_and_decimals_and_prices
+            .iter()
+            .enumerate()
+        {
+            let mut target_notional = 0i128;
+            let constituent_index = constituent_index_and_price.0;
```

Comment:
i think instead of index with like this you can do:

`for (i,[constituent_index, decimals, price]) in constituents_indexes_and_decimals_and_prices`